x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -349,30 +415,39 @@ public static Automaton buildIndexMatcherAutomaton(boolean allowRestrictedIndice
         }
 
-        private static Predicate<String> buildIndexMatcherPredicateForAction(String action, Group... groups) {
+        private static Predicate<IndexAbstraction> buildIndexMatcherPredicateForAction(String action, Group... groups) {
             final Set<String> ordinaryIndices = new HashSet<>();
             final Set<String> restrictedIndices = new HashSet<>();
+            final Set<String> grantMappingUpdatesOnIndices = new HashSet<>();
+            final Set<String> grantMappingUpdatesOnRestrictedIndices = new HashSet<>();
+            final boolean isMappingUpdateAction = action.equals(PutMappingAction.NAME) || action.equals(AutoPutMappingAction.NAME);
             for (final Group group : groups) {
                 if (group.actionMatcher.test(action)) {
                     if (group.allowRestrictedIndices) {
                         restrictedIndices.addAll(Arrays.asList(group.indices()));
                     } else {
                         ordinaryIndices.addAll(Arrays.asList(group.indices()));
                     }
+                } else if (isMappingUpdateAction &&
+                        group.privilege().name().stream().anyMatch(privilegeNameSetBwcAllowMappingUpdate::contains)) {
+                    // special BWC case for certain privileges: allow put mapping on indices and aliases (but not on data streams), even if
+                    // the privilege definition does not currently allow it
+                    if (group.allowRestrictedIndices) {
+                        grantMappingUpdatesOnRestrictedIndices.addAll(Arrays.asList(group.indices()));
+                    } else {
+                        grantMappingUpdatesOnIndices.addAll(Arrays.asList(group.indices()));
+                    }
                 }
             }
-            final Predicate<String> predicate;
-            if (restrictedIndices.isEmpty()) {
-                predicate = indexMatcher(ordinaryIndices)
-                    .and(index -> false == RestrictedIndicesNames.isRestricted(index));
-            } else if (ordinaryIndices.isEmpty()) {
-                predicate = indexMatcher(restrictedIndices);
-            } else {
-                predicate = indexMatcher(restrictedIndices)
-                    .or(indexMatcher(ordinaryIndices)
-                         .and(index -> false == RestrictedIndicesNames.isRestricted(index)));
-            }
-            return predicate;
+            final Predicate<String> namePredicate = indexMatcher(ordinaryIndices, restrictedIndices);
+            final Predicate<String> bwcSpecialCaseNamePredicate = indexMatcher(grantMappingUpdatesOnIndices,
+                    grantMappingUpdatesOnRestrictedIndices);
+            return indexAbstraction -> {
+                return namePredicate.test(indexAbstraction.getName()) ||
+                        (indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM &&
+                                (indexAbstraction.getParentDataStream() == null) &&
```

Comment:
`indexAbstraction.getParentDataStream() == null` tells if it is a concrete index that is not a backing index or an alias.
`indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM`  tells if it is a data stream.

You're suggesting that the `IndexAbstraction` have a check for whether it is a data stream or a backing index for a data stream, right?
I'm not convinced that it brings value, more often than not I prefer chaining simple tests rather than invoking a specialised test method with a long name.
I would resist adding such a test method at the moment, unless you feel strongly about it.

Reply:
I'd personally prefer to have readable method name, e.g. `IndexAbstraction#notAssociatedWithDataStream()`, over chaining logical expressions. Java method names are often long and this one does not seem to be much over it. It helps me to read higher level logic without need to figure out the details. 

I would feel strong if the expressions are repeated a few more times. In this case, it is not repeated elsewhere, so I am fine if you prefer to keep it here.